src/api/worker/facades/CalendarFacade.js

```diff
@@ -241,32 +243,41 @@ export class CalendarFacade {
 		})
 	}
 
+	/**
+	 * Load all events that have an alarm assigned.
+	 */
 	loadAlarmEvents(): Promise<Array<EventWithAlarmInfo>> {
 		const user = this._loginFacade.getLoggedInUser()
 		const alarmInfoList = user.alarmInfoList
 		if (alarmInfoList) {
 			return this._entity.loadAll(UserAlarmInfoTypeRef, alarmInfoList.alarms)
-			           .then((userAlarmInfos) =>
-				           Promise
-					           .map(userAlarmInfos, (userAlarmInfo) =>
-						           this._entity.load(CalendarEventTypeRef, [
-							           userAlarmInfo.alarmInfo.calendarRef.listId, userAlarmInfo.alarmInfo.calendarRef.elementId
-						           ])
-						               .then((event) => {
-							               return {event, userAlarmInfo}
-						               })
-						               .catch(NotFoundError, () => {
-							               // We do not allow to delete userAlarmInfos currently but when we update the server we should do that
-							               //erase(userAlarmInfo).catch(noOp)
-							               return null
-						               })
-						               .catch(NotAuthorizedError, (e) => {
-							               // Should not happen normally but could happen when user is removed from the calendar
-							               console.warn("NotAuthorized when downloading alarm event", e)
-							               return null
-						               }))
-					           .filter(Boolean) // filter out orphan alarms
-			           )
+			           .then(userAlarmInfos => {
+				           // Group referenced event ids by list id so we can load events of one list in one request.
+				           const listIdToElementIds = groupByAndMap(userAlarmInfos,
+					           (userAlarmInfo) => userAlarmInfo.alarmInfo.calendarRef.listId,
+					           (userAlarmInfo) => userAlarmInfo.alarmInfo.calendarRef.elementId)
+
+				           return promiseMap(listIdToElementIds.entries(), ([listId, elementIds]) => {
+					           return this._entity.loadMultipleEntities(CalendarEventTypeRef, listId, elementIds)
+					                      .catch(error => {
+						                      // handle NotAuthorized here because user could have been removed from group.
+						                      if (error instanceof NotAuthorizedError) {
+							                      console.warn("NotAuthorized when downloading alarm events", error)
+							                      return []
+						                      }
+						                      throw error
+					                      })
+				           }).then(flat)
```

Comment:
formatting is just borked here and for the then after and for inline condition below.

I am a bit unasy about the nested loop too, sounds like it can go very wrong.

Reply:
the find inside map you mean?

Reply:
Yes, it's effectively nested loop and the more `UserAlarmInfo's` we have the worse it gets.

Reply:
I changed it into a table lookup, also changed the return value to remove the need to pointlessly copy an event for each instance of it's alarm info. I still need to finish writing test